src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Filters/DuplicatedDependenciesSnapshotFilter.cs

```diff
@@ -33,22 +33,43 @@ internal class DuplicatedDependenciesSnapshotFilter : DependenciesSnapshotFilter
             out bool filterAnyChanges)
         {
             filterAnyChanges = false;
-            IDependency resultDependency = dependency;
+            var resultDependency = dependency;
 
-            var matchingDependency = topLevelBuilder.FirstOrDefault(
-                x => !x.Id.Equals(dependency.Id, StringComparison.OrdinalIgnoreCase)
-                     && x.ProviderType.Equals(dependency.ProviderType, StringComparison.OrdinalIgnoreCase) 
-                     && x.Caption.Equals(dependency.Caption, StringComparison.OrdinalIgnoreCase));
-            
-            // if found node with same caption, or if there were nodes with same caption but with Alias already applied
-            var shouldApplyAlias = (matchingDependency == null)
-                ? topLevelBuilder.Any(
-                    x => !x.Id.Equals(dependency.Id)
+            IDependency matchingDependency = null;
+            foreach (var x in topLevelBuilder)
+            {
+                if (!x.Id.Equals(dependency.Id, StringComparison.OrdinalIgnoreCase)
+                     && x.ProviderType.Equals(dependency.ProviderType, StringComparison.OrdinalIgnoreCase)
+                     && x.Caption.Equals(dependency.Caption, StringComparison.OrdinalIgnoreCase))
+                {
+                    matchingDependency = x;
+                    break;
+                }
+            }
+
+            // If found node with same caption, or if there were nodes with same caption but with Alias already applied
+            // NOTE: Performance sensitive, so avoid formatting the Caption with parens if it's possible to avoid it.
+            bool shouldApplyAlias;
+            if (matchingDependency == null)
+            {
+                foreach (var x in topLevelBuilder)
+                {
+                    if (!x.Id.Equals(dependency.Id)
                          && x.ProviderType.Equals(dependency.ProviderType, StringComparison.OrdinalIgnoreCase)
-                         && x.Caption.Equals(
-                             string.Format(CultureInfo.CurrentCulture, "{0} ({1})", dependency.Caption, x.OriginalItemSpec),
-                                StringComparison.OrdinalIgnoreCase))
-                : true;
+                         && x.Caption.StartsWith(dependency.Caption, StringComparison.OrdinalIgnoreCase)
+                         && string.Compare(x.Caption, dependency.Caption.Length + " (".Length, x.OriginalItemSpec, 0, x.OriginalItemSpec.Length, StringComparison.OrdinalIgnoreCase) == 0)
```

Comment:
>dependency.Caption.Length + " (".Length [](start = 54, length = 39)

potential ArgumentOutOfRangeException: item1 = MyCaption, item2 = MyCaptionX, then x.Caption.StartsWith(dependency.Caption, StringComparison.OrdinalIgnoreCase) would be true and we go to the next line which would throw

Reply:
we need to add:
var adjustedLength = dependency.Caption.Length + " (".Length;

and then use it in the condition:

x.Caption.StartsWith(dependency.Caption, StringComparison.OrdinalIgnoreCase)
&&
x.Caption.Length > adjustedLength 
&&
string.Compare(x.Caption, adjustedLength, x.OriginalItemSpec, 0, x.OriginalItemSpec.Length, StringComparison.OrdinalIgnoreCase) == 0


---
In reply to: [148041651](https://github.com/dotnet/project-system/pull/2931#discussion_r148041651) [](ancestors = 148041651)

Reply:
Addressed.